src/plugins/plugin-http.js

```diff
@@ -0,0 +1,138 @@
+}
+
+function getSpanName(requestOptions) {
+  if (isString(requestOptions)) {
+    requestOptions = url.parse(requestOptions);
+  }
+  // c.f. _http_client.js ClientRequest constructor
+  return requestOptions.hostname || requestOptions.host || 'localhost';
+}
+
+function setTraceHeader(parsedOptions, context, api) {
+  if (context) {
+    return merge(parsedOptions, {
+      headers: {
+        [constants.TRACE_CONTEXT_HEADER_NAME]: context
+      }
+    });
+  }
+  return parsedOptions;
+}
+
+function extractUrl(parsedOptions) {
+  var uri = parsedOptions;
+  var agent = parsedOptions._defaultAgent || httpAgent.globalAgent;
+  return isString(uri) ? uri :
+    (parsedOptions.protocol || agent.protocol) + '//' +
+    (parsedOptions.hostname || parsedOptions.host || 'localhost') +
+    ((isString(parsedOptions.port) ? (':' + parsedOptions.port) : '')) +
+    (parsedOptions.path || parsedOptions.pathName || '/');
+}
+
+function parseRequestOptions(requestOptions) {
+  return isString(requestOptions) ?
+    merge({headers: {}}, url.parse(requestOptions)) :
+    merge({headers: {}}, requestOptions);
+}
+
+function patchedHTTPRequest(requestOptions, callback, request, api) {
+  if (!requestOptions) {
+    return request.call(request, requestOptions, callback);
+  } else if (isTraceAgentRequest(requestOptions)) {
+    return request.call(request, requestOptions, callback);
+  } else if (!api.getRootSpan()) {
+    if (isString(requestOptions)) {
+      requestOptions = url.parse(requestOptions);
+    }
+    // What's the new logger target?
+    // console.log('Untraced http uri:', extractUrl(requestOptions));
+    return request.call(request, requestOptions, callback);
+  }
+  var parsedOptions = parseRequestOptions(requestOptions);
+  var uri = extractUrl(parsedOptions);
+  var requestLifecycleSpan = api.createChildSpan({name: getSpanName(parsedOptions)});
+  if (!requestLifecycleSpan) {
+    // Bail out since we couldn't get a span
+    return request.call(request, requestOptions, callback);
+  }
+  requestLifecycleSpan.addLabel(api.labels.HTTP_METHOD_LABEL_KEY,
+    parsedOptions.method || 'GET');
+  requestLifecycleSpan.addLabel(api.labels.HTTP_URL_LABEL_KEY, uri);
+  parsedOptions = setTraceHeader(parsedOptions, requestLifecycleSpan.getTraceContext(), api);
+  var req = request.call(request, parsedOptions, function (res) {
+    api.wrapEmitter(res);
+    var numBytes = 0;
+    res.on('data', function (chunk) {
+      numBytes += chunk.length;
+    });
+    res.on('end', function () {
+      if (requestLifecycleSpan) {
+        requestLifecycleSpan
+          .addLabel(api.labels.HTTP_RESPONSE_SIZE_LABEL_KEY, numBytes);
+        requestLifecycleSpan
+          .addLabel(api.labels.HTTP_RESPONSE_CODE_LABEL_KEY, res.statusCode);
+        requestLifecycleSpan.endSpan();
+      }
+    });
+    if (callback) {
+      return callback(res);
+    }
+  });
+  api.wrapEmitter(req);
+  req.on('error', function (e) {
+    if (e && requestLifecycleSpan) {
```

Comment:
Same as above; I don't think `requestLifecycleSpan`'s existence needs to be checked here.